File: RentalModule.Application/CommandHandlers/CarOffer/CreateCarOffer/CreateCarOfferCommand.cs
using System.Collections.Immutable;
using System.ComponentModel.DataAnnotations;
using RentalModule.Application.Contract.CarOffers.CreateCarOffer;
using Results.Domain;
using Shared.CQRS;
using Shared.CQRS.CommandHandlers;

namespace RentalModule.Application.CommandHandlers.CarOffer.CreateCarOffer;

public class CreateCarOfferCommand : ICommand<HandlerResult<CreateCarOfferResponse, IErrorResult>>
{
    [StringLength(200)]
    [Required]
    public string Heading { get; init; } = string.Empty;

    [StringLength(500)]
    [Required]
    public string ShortDescription { get; init; } = string.Empty;

    [Url]
    [Required]
    public string FeaturedImageUrl { get; init; } = string.Empty;

    [RegularExpression(@"^[a-zA-Z0-9\-]+$", ErrorMessage = "UrlHandle must be alphanumeric with hyphens allowed.")]
    [StringLength(100)]
    [Required]
    public string UrlHandle { get; init; } = string.Empty;

    [StringLength(50)]
    [Required]
    public string Horsepower { get; init; } = string.Empty;

    [Range(1886, 2100)]
    [Required]
    public int YearOfProduction { get; init; }

    [StringLength(20)]
    [Required]
    public string EngineDetails { get; init; } = string.Empty;

    [StringLength(10)]
    [Required]
    public string DriveDetails { get; init; } = string.Empty;

    [StringLength(10)]
    [Required]
    public string GearboxDetails { get; init; } = string.Empty;

    [StringLength(50)]
    [Required]
    public string CarDeliverylocation { get; init; } = string.Empty;

    [StringLength(50)]
    [Required]
    public string CarReturnLocation { get; init; } = string.Empty;

    [Required]
    public DateTime PublishedDate { get; init; } = DateTime.Today;

    [Required]
    public bool Visible { get; init; }

    [Range(0, double.MaxValue)]
    [Required]
    public double OneNormalDayPrice { get; init; }

    [Range(0, double.MaxValue)]
    [Required]
    public double OneWeekendDayPrice { get; init; }

    [Range(0, double.MaxValue)]
    [Required]
    public double FullWeekendPrice { get; init; }

    [Range(0, double.MaxValue)]
    [Required]
    public double OneWeekPrice { get; init; }

    [Range(0, double.MaxValue)]
    [Required]
    public double OneMonthPrice { get; init; }
    
    public ImmutableArray<string>? Tags { get; init; }

    //[Urls]todo
    public ImmutableArray<string>? ImageUrls { get; init; }

}

File: RentalApi/Application/command_handlers/car_offer/create_car_offer/create_car_offer_command.go
package commands

import (
	models "rental-api/Domain/models/external"
)

type CreateCarOfferCommand struct {
	Heading            string   `json:"heading" binding:"required" example:"Car Offer Title" swaggertype:"string" validate:"required"`
	ShortDescription   string   `json:"shortDescription" binding:"required" example:"Short description" swaggertype:"string" validate:"required"`
	FeaturedImageUrl   string   `json:"featuredImageUrl" example:"https://example.com/image.jpg" swaggertype:"string"`
	UrlHandle          string   `json:"urlHandle" binding:"required" example:"car-offer-title" swaggertype:"string" validate:"required"`
	Horsepower         string   `json:"horsepower" example:"300" swaggertype:"string"`
	YearOfProduction   int      `json:"yearOfProduction" example:"2023" swaggertype:"integer"`
	EngineDetails      string   `json:"engineDetails" example:"2.0L Turbo" swaggertype:"string"`
	DriveDetails       string   `json:"driveDetails" example:"AWD" swaggertype:"string"`
	GearboxDetails     string   `json:"gearboxDetails" example:"Automatic" swaggertype:"string"`
	Visible            bool     `json:"visible" example:"true" swaggertype:"boolean"`
	OneNormalDayPrice  float64  `json:"oneNormalDayPrice" example:"100.00" swaggertype:"number"`
	OneWeekendDayPrice float64  `json:"oneWeekendDayPrice" example:"150.00" swaggertype:"number"`
	OneWeekPrice       float64  `json:"oneWeekPrice" example:"600.00" swaggertype:"number"`
	OneMonthPrice      float64  `json:"oneMonthPrice" example:"2000.00" swaggertype:"number"`
	Tags               []string `json:"tags" example:"[\"luxury\",\"sports\"]" swaggertype:"array,string"`
	ImageUrls          []string `json:"imageUrls" example:"[\"https://example.com/image1.jpg\"]" swaggertype:"array,string"`
	models.JwtToken    `json:",inline"`
}
